programs/merkle-distributor/src/instructions/claim_and_stake/new_claim_and_stake.rs

```diff
@@ -0,0 +1,212 @@
+use anchor_lang::{
+    context::Context, prelude::*, solana_program::hash::hashv, system_program::System, Accounts,
+    Key, Result,
+};
+use anchor_spl::token::{Token, TokenAccount};
+use jito_merkle_verify::verify;
+
+use crate::{
+    error::ErrorCode,
+    math::SafeMath,
+    state::{
+        claim_status::ClaimStatus, claimed_event::NewClaimEvent,
+        merkle_distributor::MerkleDistributor,
+    },
+};
+
+use locked_voter::program::LockedVoter as Voter;
+use locked_voter::{self as voter, Escrow, Locker};
+
+// We need to discern between leaf and intermediate nodes to prevent trivial second
+// pre-image attacks.
+// https://flawed.net.nz/2018/02/21/attacking-merkle-trees-with-a-second-preimage-attack
+const LEAF_PREFIX: &[u8] = &[0];
+
+/// [merkle_distributor::new_claim_and_stake] accounts.
+#[derive(Accounts)]
+pub struct NewClaimAndStake<'info> {
+    /// The [MerkleDistributor].
+    #[account(mut, has_one = locker)]
+    pub distributor: AccountLoader<'info, MerkleDistributor>,
+
+    /// Claim status PDA
+    #[account(
+        init,
+        seeds = [
+            b"ClaimStatus".as_ref(),
+            claimant.key().to_bytes().as_ref(),
+            distributor.key().to_bytes().as_ref()
+        ],
+        bump,
+        space = 8 + ClaimStatus::LEN,
+        payer = claimant
+    )]
+    pub claim_status: Account<'info, ClaimStatus>,
+
+    /// Distributor ATA containing the tokens to distribute.
+    #[account(
+        mut,
+        associated_token::mint = distributor.load()?.mint,
+        associated_token::authority = distributor.key(),
+        address = distributor.load()?.token_vault
+    )]
+    pub from: Account<'info, TokenAccount>,
+
+    /// Who is claiming the tokens.
+    #[account(mut)]
+    pub claimant: Signer<'info>,
+
+    /// operator
+    pub operator: Option<Signer<'info>>,
+
+    /// SPL [Token] program.
+    pub token_program: Program<'info, Token>,
+
+    /// The [System] program.
+    pub system_program: Program<'info, System>,
+
+    /// Voter program
+    pub voter_program: Program<'info, Voter>,
+
+    /// CHECK: Locker
+    #[account(mut)]
+    pub locker: Box<Account<'info, Locker>>,
```

Comment:
It seems for the permissionless variant, the claimer might claim into the wrong locker?

Reply:
there is a check in line 29 `#[account(mut, has_one = locker)]`